backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportHeaderGenerator.kt

```diff
@@ -599,7 +633,43 @@ internal class ObjCExportTranslatorImpl(
         return if (classDescriptor.isInterface) {
             ObjCProtocolType(referenceProtocol(classDescriptor).objCName)
         } else {
-            ObjCClassType(referenceClass(classDescriptor).objCName)
+            val typeArgs = if (writeGenerics) {
+                kotlinType.arguments.map { typeProjection ->
+                    mapReferenceTypeIgnoringNullability(typeProjection.type, typeParamProvider)
+                }
+            } else {
+                emptyList()
+            }
+            ObjCClassType(referenceClass(classDescriptor).objCName, typeArgs)
+        }
+    }
+
+    internal fun ClassDescriptor.genericParamDeclaration(paramProvider: TypeParamProvider, includeUpperBounds:Boolean = false): List<String> = declaredTypeParameters.map { typeParam ->
+        val param = "${typeParam.variance.objcDeclaration()}${typeParam.name.asString()}"
+
+        if(includeUpperBounds) {
+            val upperBounds = typeParam.upperBounds
+            val upperBoundTypes = upperBounds
+                    .filter { kt ->
+                        val classDescriptor = kt.getErasedTypeClass()
+                        !(classDescriptor == builtIns.any || classDescriptor.classId in mapper.hiddenTypes || classDescriptor.isInlined() || classDescriptor.defaultType.isObjCObjectType())
```

Comment:
Not very readable condition.

Reply:
I'm pulling this out. The upper bounds constraint on generics was one of the last things added and was disabled due to some complicated types. If reintroduced this logic will need to be resurrected in some form, but it's in commit history for reference.